src/backend/distributed/commands/seclabel.c

```diff
@@ -0,0 +1,165 @@
+/*-------------------------------------------------------------------------
+ *
+ * seclabel.c
+ *
+ *    This file contains the logic of SECURITY LABEL statement propagation.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "commands/seclabel.h"
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/coordinator_protocol.h"
+#include "distributed/deparser.h"
+#include "distributed/log_utils.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/metadata/distobject.h"
+
+
+PG_FUNCTION_INFO_V1(citus_test_register_label_provider);
+
+
+/*
+ * citus_test_register_label_provider registers a dummy label provider
+ * named 'citus_tests_label_provider'. This is aimed to be used for testing.
+ */
+Datum
+citus_test_register_label_provider(PG_FUNCTION_ARGS)
+{
+	register_label_provider("citus_tests_label_provider", citus_test_object_relabel);
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * PreprocessSecLabelStmt is executed before the statement is applied to the local
+ * postgres instance.
+ *
+ * In this stage we can prepare the commands that need to be run on all workers to assign
+ * security labels on distributed objects, currently supporting just Role objects.
+ */
+List *
+PreprocessSecLabelStmt(Node *node, const char *queryString,
+					   ProcessUtilityContext processUtilityContext)
+{
+	if (!IsCoordinator() || !ShouldPropagate())
```

Comment:
Is there a reason to not incorporate the logic here into Postprocess too?

Reply:
If there is a reason, should this be changed to following as we do in Postprocess?
```suggestion
	if (!EnableCreateRolePropagation || !IsCoordinator() || !ShouldPropagate())
```

Reply:
Hmm, I see your point. So, in Preprocess, I would also like to give the `NOTICE` for not propagating the command for distributed objects that are not Roles. Hence, the Preprocess is different. In Postprocess, however, I don't intend to do anything for distributed objects that are not Roles.

Reply:
I think this still doesn't prevent having two logics in Postprocess. In general, having prev_ProcessUtility first then kicking Citus logic feels safer to me because that way we can rely on Postgres checking the sanity of stmt object already. Otherwise, e.g., deparser might need to care about unexpected nulls / incorrect values etc. in the stmt object.

Reply:
Hmm, so you suggest incorporate the logic from Preprocess into Postprocess? I guess I agree with that.